### ISSUE TITLE
[BUG] MoE router can potentially hack the aux loss when using sigmoid gating function

### ISSUE THREAD

**Issue by epwalsh** [NVIDIA/Megatron-LM#1545]
**Describe the bug**
It's possible for the MoE router to "hack" the auxiliary load balancing loss by pushing sigmoid activations below 1e-20. While I haven't seen this happen in Megatron-LM yet, I have seen it in https://github.com/allenai/OLMo-core/pull/255 when we used the same normalizer epsilon of 1e-20, so I thought I'd raise this here just in case this happens to someone else.

https://github.com/NVIDIA/Megatron-LM/blob/55c968dabb6d32cf8c21b1e3fadeea47e2ceb78d/megatron/core/transformer/moe/router.py#L195-L199

**To Reproduce**
Train an MoE with auxiliary load balancing loss and the sigmoid router gating function. It's rather hard to reproduce as it doesn't happen all the time. The model needs to get somewhat lucky to "realize" that it can hack the aux loss.

**Expected behavior**
Aux loss stays >= 1.

**Environment (please complete the following information):**
 - 55c968dabb6d32cf8c21b1e3fadeea47e2ceb78d

**Proposed fix**
Decrease the normalizing epsilon (e.g. use `torch.finfo(torch.float32).tiny` ~= `1.175e-38`) or remove it. I'm not sure it's necessary for stability.

**Comment by Victarry** [comment]
Thanks for your feedback! We will take a look.

**Comment by Skylion007** [comment]
Okay, this is even more insidious than this issue suggests. If you are really unlucky with expert bias, it can initialize near 0 too which makes it very hard to detect.

**Comment by epwalsh** [comment]
And this loss function is actually fundamentally flawed as @2015aroras discovered. It can be hacked even without exploiting floating point precision issues. In the [Switch Transformers paper](https://arxiv.org/pdf/2101.03961) (which I believe was the first paper to introduce this exact form of the LBL function?), they say the loss is minimized at 0 with uniform load balancing, but that's not true.

Here's a simple example with 2 experts that gives a loss of 0.875 with very imbalanced assignment:

```python
import torch

from olmo_core.ops import moe as ops

# NOTE: ignoring batch dimension so instance-level LBL is the same as micro-batch-level LBL.

def load_balancing_loss(scores: torch.Tensor, batch_size_per_expert: torch.Tensor) -> torch.Tensor:
    num_experts = scores.shape[-1]
    scores_fraction = scores.mean(dim=0)
    batch_size_per_expert_fraction = batch_size_per_expert / batch_size_per_expert.sum()
    return num_experts * (scores_fraction * batch_size_per_expert_fraction).sum()


def compute_losses(expert_scores: torch.Tensor):
    num_experts = expert_scores.shape[-1]
    assert (expert_scores.sum(dim=-1) == 1.0).all()
    _, expert_indices = torch.topk(expert_scores, 1)
    batch_size_per_expert = ops.batched_histc(expert_indices, num_experts).sum(dim=0)
    print(f"LBL    = {load_balancing_loss(expert_scores, batch_size_per_expert):.4f}")


delta = 1e-5

print("Hacked case:")
expert_scores = torch.tensor(
    [
        [1.00 - delta, 0.00 + delta],
        [1.00 - delta, 0.00 + delta],
        [0.50 - delta, 0.50 + delta],
        [0.50 - delta, 0.50 + delta],
        [0.50 - delta, 0.50 + delta],
        [0.50 - delta, 0.50 + delta],
        [0.50 - delta, 0.50 + delta],
        [0.50 - delta, 0.50 + delta],
    ]
)
compute_losses(expert_scores)
```

One thing that could help here would be to optimize for `(LBL - 1.0)^2`.

**Comment by Victarry** [comment]
cc @yanring @aklife97

**Comment by abhinav-khattar** [comment]
Thanks for reporting this @epwalsh!

> It's possible for the MoE router to "hack" the auxiliary load balancing loss by pushing sigmoid activations below 1e-20

That is right, but we haven't practically ever seen this happen on a huge number of MoE runs. What probably stops this from happening is that if `P_i ` in the aux-loss term goes too low, so does the probability value that the expert output is multiplied with. That is, if `P_i -> 0` the MoE layer output will likely 0 out as well. While it is theoretically possible to come up with cases where activation is very slightly less than 1e-20 and expert does not 0 out, it should be very rare (as you mentioned) especially if aux-loss-coeff is not extremely high. 

The aux-loss curves you pasted from the OLMo repository look intriguing. We haven't seen that behavior, and I would suspect the aux-loss is likely over-powering the total loss in that scenario.



> use torch.finfo(torch.float32).tiny

Great suggestion, we can make this change to ensure the above case is even more rare. But again, there is still no guarantee that the model would not take the sigmoid activation all the way to 0 to reduce aux-loss if aux-loss were to overpower the total loss.



> And this loss function is actually fundamentally flawed

Yes... We have internally discussed cases where the loss goes to 1 with fully imbalanced load, something like below:

```
[
     [0.5 - eps, 0.5 + eps],
     [0.5 - eps, 0.5 + eps]
]
```

But its great to know that 1 is not the minima of the loss either (as has been claimed) with the case you presented.



> One thing that could help here would be to optimize for (LBL - 1.0)^2

Thanks for the recommendation, but this formulation still has its shortcomings. For instance, in the case above, this new loss still fails and 0s out at perfectly imbalanced load.

**Comment by epwalsh** [comment]
> The aux-loss curves you pasted from the OLMo repository look intriguing. We haven't seen that behavior, and I would suspect the aux-loss is likely over-powering the total loss in that scenario.

This is true! We've found that as long as we keep the scaled LBL loss relatively small compared to CE loss, we don't see this degenerate behavior.
